File: src/ros1_for_reference/dofbot_moveit/src/02_motion_plan.cpp
#include <iostream>
#include "ros/ros.h"
#include <moveit/move_group_interface/move_group_interface.h>
#include <tf/LinearMath/Quaternion.h>

using namespace std;
// 角度转弧度
const float DE2RA = M_PI / 180.0f;

int main(int argc, char **argv) {
    ros::init(argc, argv, "dofbot_motion_plan_cpp");
    ros::NodeHandle n;
    // 设置线程  set thread
    ros::AsyncSpinner spinner(1);
    // 开启线程  open thread
    spinner.start();
    // Initialize the robotic arm motion planning group
    // 初始化机械臂运动规划组
    moveit::planning_interface::MoveGroupInterface dofbot("dofbot");
    dofbot.allowReplanning(true);
    dofbot.setPlanningTime(5);
    dofbot.setNumPlanningAttempts(10);
    dofbot.setGoalPositionTolerance(0.01);
    dofbot.setGoalOrientationTolerance(0.01);
    dofbot.setMaxVelocityScalingFactor(1.0);
    dofbot.setMaxAccelerationScalingFactor(1.0);
    dofbot.setNamedTarget("down");
    dofbot.move();
    sleep(0.5);
    geometry_msgs::Pose pose;
    pose.position.x = 0.0;
    pose.position.y = 0.05957241;
    pose.position.z = 0.1680498;
    tf::Quaternion quaternion;
    // RPY的单位是角度值  The unit of RPY is the angle value
    double Roll = -140;
    double Pitch = 0.0;
    double Yaw = 0.0;
    // RPY转四元数  RPY to Quaternion
    quaternion.setRPY(Roll * DE2RA, Pitch * DE2RA, Yaw * DE2RA);
//    pose.orientation.x = 0.940132;
//    pose.orientation.y = -0.000217502;
//    pose.orientation.z = 0.000375234;
//    pose.orientation.w = -0.340811;
    pose.orientation.x = quaternion.x();
    pose.orientation.y = quaternion.y();
    pose.orientation.z = quaternion.z();
    pose.orientation.w = quaternion.w();
    string link = dofbot.getEndEffectorLink();
    // 设置目标点  set target point
    dofbot.setPoseTarget(pose, link);
    int index = 0;
    // 多次执行,提高成功率  Execute multiple times to improve the success rate
    while (index <= 10) {
        moveit::planning_interface::MoveGroupInterface::Plan plan;
        // 运动规划  motion planning
        const moveit::planning_interface::MoveItErrorCode &code = dofbot.plan(plan);
        if (code == code.SUCCESS) {
            ROS_INFO_STREAM("plan success");
            dofbot.execute(plan);
            break;
        } else {
            ROS_INFO_STREAM("plan error");
        }
        index++;
    }
    return 0;
}

File: src/ros1_for_reference/dofbot_moveit/scripts/02_motion_plan.py
#!/usr/bin/env python
# coding: utf-8
import rospy
from math import pi
from time import sleep
from geometry_msgs.msg import Pose
from moveit_commander.move_group import MoveGroupCommander
from tf.transformations import quaternion_from_euler

# Convert radians to degrees
# 弧度转换成角度
DE2RA = pi / 180

if __name__ == '__main__':
    rospy.init_node("dofbot_motion_plan_py")
    # Initialize the robotic arm motion planning group
    # 初始化机械臂运动规划组
    dofbot = MoveGroupCommander("dofbot")
    # Allow replanning when motion planning fails
    # 当运动规划失败后，允许重新规划
    dofbot.allow_replanning(True)
    dofbot.set_planning_time(5)
    # number of attempts to plan
    # 尝试规划的次数
    dofbot.set_num_planning_attempts(10)
    # Set allowable target position error
    # 设置允许目标位置误差
    dofbot.set_goal_position_tolerance(0.01)
    # Set the allowable target attitude error
    # 设置允许目标姿态误差
    dofbot.set_goal_orientation_tolerance(0.01)
    # Set allowable target error
    # 设置允许目标误差
    dofbot.set_goal_tolerance(0.01)
    # set maximum speed
    # 设置最大速度
    dofbot.set_max_velocity_scaling_factor(1.0)
    # set maximum acceleration
    # 设置最大加速度
    dofbot.set_max_acceleration_scaling_factor(1.0)
    # 设置"down"为目标点
    dofbot.set_named_target("down")
    dofbot.go()
    sleep(0.5)
    # Create a pose instance
    # 创建位姿实例
    pos = Pose()
    # Set a specific location
    # 设置具体的位置
    pos.position.x = 0.0
    pos.position.y = 0.0597016
    pos.position.z = 0.168051
    # The unit of RPY is the angle value
    # RPY的单位是角度值
    roll = -140.0
    pitch = 0.0
    yaw = 0.0
    # RPY to Quaternion
    # RPY转四元数
    q = quaternion_from_euler(roll * DE2RA, pitch * DE2RA, yaw * DE2RA)
    # pos.orientation.x = 0.940132
    # pos.orientation.y = -0.000217502
    # pos.orientation.z = 0.000375234
    # pos.orientation.w = -0.340811
    pos.orientation.x = q[0]
    pos.orientation.y = q[1]
    pos.orientation.z = q[2]
    pos.orientation.w = q[3]
    # set target point
    # 设置目标点
    dofbot.set_pose_target(pos)
    # Execute multiple times to improve the success rate
    # 多次执行,提高成功率
    for i in range(5):
        # motion planning
        # 运动规划
        plan = dofbot.plan()
        if len(plan.joint_trajectory.points) != 0:
            print ("plan success")
            # Run after planning is successful
            # 规划成功后运行
            dofbot.execute(plan)
            break
        else:
            print ("plan error")
